### PR TITLE
Fixed #17657 -- Made ModelForm respect ModelMultipleChoiceField's to_field_name.

### PR DESCRIPTION
**Pull request by andreyfedoseev**
…tead of a list of pks, let the form fields handle it.

Remove obsolete tests for `model_to_dict` performance optimization.

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/model_forms/tests.py [hunk 1]
```diff
@@ -1789,6 +1717,49 @@ class WriterForm(forms.Form):
...
+
+            class Meta:
+                model = Category
+                fields = ["categories"]
+
+        Category.objects.create(name='Category 1', slug='category-1')
+        category_2 = Category.objects.create(name='Category 2', slug='category-2')
+        category_3 = Category.objects.create(name='Category 3', slug='category-3')
+
+        article = Article(
```
Comment (timgraham) [inline tests/model_forms/tests.py]: `.object.create`

File: tests/model_forms/tests.py [hunk 2]
```diff
@@ -1789,6 +1717,49 @@ class WriterForm(forms.Form):
...
+    def test_to_field_name_17657(self):
+        """
+        #17657 -- Make sure ModelForm respects to_field_name attribute of ModelMultipleChoiceField
+        """
+
+        class ArticleForm(forms.ModelForm):
+
+            categories = forms.ModelMultipleChoiceField(
+                Category.objects.all(),
+                to_field_name="slug"
```
Comment (timgraham) [inline tests/model_forms/tests.py]: use single quotes

File: tests/model_forms/tests.py [hunk 3]
```diff
@@ -1789,6 +1717,49 @@ class WriterForm(forms.Form):
...
 
+class MultipleChoiceWithFieldNameTests(TestCase):
+
+    def test_to_field_name_17657(self):
+        """
+        #17657 -- Make sure ModelForm respects to_field_name attribute of ModelMultipleChoiceField
+        """
+
+        class ArticleForm(forms.ModelForm):
+
```
Comment (timgraham) [inline tests/model_forms/tests.py]: chop blank line

File: tests/model_forms/tests.py [hunk 4]
```diff
@@ -1789,6 +1717,49 @@ class WriterForm(forms.Form):
         self.assertEqual(len(params), 1)
 
 
+class MultipleChoiceWithFieldNameTests(TestCase):
```
Comment (timgraham) [inline tests/model_forms/tests.py]: The class name seems a bit specific -- not sure any other tests would fit in this class (the existing organization is a bit poor).

File: tests/model_forms/tests.py [hunk 5]
```diff
@@ -1789,6 +1717,49 @@ class WriterForm(forms.Form):
...
+
+    def test_to_field_name_17657(self):
```
Comment (timgraham) [inline tests/model_forms/tests.py]: chop ticket number

File: tests/model_forms/tests.py [hunk 6]
```diff
@@ -1789,6 +1717,49 @@ class WriterForm(forms.Form):
...
+        """
+        #17657 -- Make sure ModelForm respects to_field_name attribute of ModelMultipleChoiceField
```
Comment (timgraham) [inline tests/model_forms/tests.py]: chop "Make sure" in favor of simply stating the expected behavior. The ticket reference isn't really needed -- we reserve that for obscure issues.

File: tests/model_forms/tests.py [hunk 7]
```diff
@@ -1789,6 +1717,49 @@ class WriterForm(forms.Form):
...
+        """
+
+        class ArticleForm(forms.ModelForm):
+
+            categories = forms.ModelMultipleChoiceField(
```
Comment (timgraham) [inline tests/model_forms/tests.py]: a single line is fine here (up to 119 chars is okay if it helps readability)

File: tests/model_forms/tests.py [hunk 8]
```diff
@@ -1789,6 +1717,49 @@ class WriterForm(forms.Form):
...
+
+        article = Article(
+            headline='Test article',
+            slug='test-article',
+            pub_date=datetime.date(1988, 1, 4),
+            writer=Writer.objects.create(name='Test writer'),
+            article='Hello.'
+        )
+        article.save()
+
```
Comment (timgraham) [inline tests/model_forms/tests.py]: chop blank lines here and below

File: tests/model_forms/tests.py [hunk 9]
```diff
@@ -1789,6 +1717,49 @@ class WriterForm(forms.Form):
...
+        article.categories.add(category_2, category_3)
+
+        form = ArticleForm(instance=article)
+
+        self.assertHTMLEqual(
+            str(form["categories"]), """<select id="id_categories" multiple="multiple" name="categories" required>
```
Comment (timgraham) [inline tests/model_forms/tests.py]: move the string to the next line as done in other tests in this file

File: django/db/models/fields/related.py [hunk 10]
```diff
@@ -1662,12 +1662,7 @@ def value_from_object(self, obj):
         """
         Return the value of this field in the given model instance.
         """
-        if obj.pk is None:
```
Comment (timgraham) [inline django/db/models/fields/related.py]: I think we'll need to keep the `if obj.pk is None` condition to fix the test failure.

**Comment by andreyfedoseev** [comment]
@timgraham Thank you for your feedback. Do you want to me squash all the fix-ups to a single commit?

**Comment by timgraham** [comment]
Yes, that's the normal procedure.

**Comment by andreyfedoseev** [comment]
I updated the pull request, please review.

I moved my test to `ModelMultipleChoiceFieldTests` and refactored it to remove unnecessary definitions of categories. It doesn't rely on checking the HTML output of the form widget, so it's less fragile now.

**Comment by timgraham** [comment]
Code looks good but now do we have to reopen [#18949](https://code.djangoproject.com/ticket/18949) since we removed that optimization? It's not obvious to me what "let the form fields handle it" means.

**Comment by andreyfedoseev** [comment]
> Code looks good but now do we have to reopen #18949 since we removed that optimization?

I removed not only the optimization, but all the code for the special handling of M2M fields in `model_to_dict`. Let me explain:

`model_to_dict` used to have this special case for M2M fields. `M2MField.value_from_object` returned a `QuerySet` and `model_to_dict` converted it to a list of primary keys. This special case:

1. Caused a bug, see https://code.djangoproject.com/ticket/17657
2. Was inefficient (thus a number of optimization has followed)
3. Was unnecessary. Removing this special case doesn't seem to break anything. On the contrary, removing it fixes the bug mentioned in (1). We use a patched version of Django with this special case removed in production for about 4 years now.

Here's what I mean when I say "let the form field handle it": M2M field value returned from `model_to_dict` is passed to `ModelMultipleChoiceField.prepare_value`. This method can accept both a `QuerySet` or a list of primary key (or other field's values if `to_field_name` is used). Although, if `to_field_name` attribute is used, `prepare_value` will return correct result only if a `QuerySet` is passed as a value. See `ModelMultipleChoiceField.prepare_value` and `ModelChoiceField.prepare_value` for details.

**Comment by timgraham** [comment]
I suppose correct behavior should be the first priority. Merged in 81963b37a92ef583b9f725f1a78dd2ca97c1ab95, thanks!

**Pull request closed without merging**